### ISSUE TITLE
implementing String.__mod__

### ISSUE THREAD

**Issue by jstoebel** [pybee/batavia#247]
The modulus operator for Str seems straight forward enough. for example:

```
>>> 'spam' % b'123'
'spam'
```

However, someone very smart wrote the tests here:

```
>>> x = "One arg: %s"
>>> y = slice(5)
>>> x % y
One arg: slice(None, 5, None)
```

The mod operator is misunderstood as string formatting. Any guidance for how to implement this? I could use a regex to check if the left operand has a string formatting marker in it, but there might be a smarter method. I welcome input.

**Comment by freakboy3742** [comment]
This is a bit of a hairy one. The mod operator on String is one of the most complex operators, because under the hood it's Python's string substitution, which has a lot of edge cases. 

There's [an existing utility method](https://github.com/pybee/batavia/blob/master/batavia/utils.js#L112) that does *part* of the job; but there's obviously a lot of type handling and special cases that need to be handled.

**Comment by wolfv** [comment]
Brython also seems to have a fairly complete implementation for reference, maybe? I can't judge how good it is: https://github.com/brython-dev/brython/blob/63b2c5cacf3d15aca20f40cb16986e1bc65ee18c/www/src/py_string.js#L580

**Comment by freakboy3742** [comment]
@wolfv I've only had a quick look, but it looks like a pretty good implementation, and the Brython license allows re-use with attribution. It may well be worth borrowing...

**Comment by jstoebel** [comment]
thanks. I'd like to work on this if that's alright.

**Comment by jstoebel** [comment]
what is the right way to credit Brython's implementation? 
https://github.com/brython-dev/brython/blob/63b2c5cacf3d15aca20f40cb16986e1bc65ee18c/LICENCE.txt

**Comment by jstoebel** [comment]
I've started work on porting over the function from Brython and have hit an interesting snag. Here is the [commit:](https://github.com/jstoebel/batavia/commit/4414f8185a51b0778475a25b79be754b1f2f35ca#diff-738bea61621ebe472834e9b3a17930aaR417)

I added functionality at the top of the function to detect if the number of arguments given != the number of format markers and throw the correct error if so. Now I am getting different results in the test compared with trying it in the browser. For example given the input:

```
>>> x = "Three args: %s | %s | %s"
>>> y = True
>>> x % y
```
trying in my browser gives the correct result: `TypeError: not enough arguments for format string`

However running the test `tests.datatypes.test_str.BinaryStrOperationTests.test_modulo_bool`
gives a different result:

```
  >>> x = "Three args: %s | %s | %s"
  >>> y = True
  >>> x % y
- <class 'TypeError'> : not all arguments converted during string formatting
+ <class 'TypeError'> : not enough arguments for format string
```

Any insight as to what is going wrong here? I brought a good deal of code in from Brython and can't pretend to understand all of it, but its strange that trying in the browser seems to behave differently than when running the test. What am I missing?

**Comment by swenson** [comment]
The tests run using PhantomJS. I wonder if there is some subtle difference in the way they handle regexes?

**Comment by jstoebel** [comment]
thanks I will investigate that.

**Comment by jstoebel** [comment]
Firing up the phantomjs CLI I find that this rexep does not work

```
/\x25(?:([1-9]\d*)\$|\(([^\)]+)\))?(\+)?(0|'[^$])?(-)?(\d+)?(?:\.(\d+))?([b-gijosuxX])/g
``` 
I get the error: `Invalid regular expression: nothing to repeat` even though it does compile in node.

Through some trial and error I found that removing the `(\+)?` causes the problem to go away. Strangely, phantomjs does not even like the simple regex `(\+)` (throws the same error)

I filed a [ticket](https://github.com/ariya/phantomjs/issues/14525) with phantomjs, but is there something I a missing here?
